lldb/source/Plugins/SymbolFile/DWARF/DWARFDeclContext.h

```diff
@@ -68,6 +68,11 @@ class DWARFDeclContext {
 
   const char *GetQualifiedName() const;
 
+  /// Returns a vector of string, one string per entry in the fully qualified
+  /// name. For example, for the DeclContext `A::B::C`, this methods returns
+  /// `{"C", "B", "A"}`
+  llvm::SmallVector<llvm::StringRef> GetQualifiedNameAsVector() const;
```

Comment:
Would it make sense to use CompilerContext (which is a vector of `pair<nodeKind, string>`) instead?

Reply:
I'm not sure what the relationship between a CompilerContext and a DWARFDeclContext (which is what we have here) is. Let me try to figure out if there is some way to connect the two

Reply:
AFAICT DWARFDeclContext has no existing relationship to CompilerContexts.

Even if we were to add one, I think we would lose some information here. DWARFDeclContext keeps a `dwarf::Tag` instead of the more specific `CompilerContextKind` enumeration, which is more of a bitfield for specifying queries (based on the comments of that class). 

... and CompilerContexts use ConstStrings 🙄

Reply:
WFM

Reply:
This does seem like a duplication of a `std::vector<CompilerContext>`, so I would vote to possibly switch to using `std::vector<CompilerContext>` as this is what is used for type lookups to do the matching. If anything in this name match can actually be anything, the `CompilerContext::kind` can be set to `CompilerContextKind::Any` easily.

Reply:
One idea would be to get rid of `DWARFDeclContext` and just use `std::vector<CompilerContext>` everywhere. 

We could make a using directive for `std::vector<CompilerContext>` so we don't have to mention the std::vector all of the time :
```
using CompilerDeclContext= std::vector<CompilerContext>;
```
Then we could actually get rid of `DWARFDeclContext` and just use `CompilerDeclContext` everywhere, as it really contains very similar kind of stuff, it is just more DWARF specific. We already make `ConstString` objects for each DIE's name anyway and the comparing many matching entries becomes a pointer compare for the string instead of actual string comparisons.

Reply:
Out of curiosity, why is a pair of name + tag called a "Compiler Context"?

> We could make a using directive for std::vector<CompilerContext> so we don't have to mention the std::vector all of the time :

A DWARFDeclContext has a couple of functions that are used elsewhere, so we can't replace it with a vector 1:1.  We'd probably still need an abstraction or a few free functions. 
Also [personal opinion] it is better to have `std::vector` if the abstraction _is_ just vector.

> comparing many matching entries becomes a pointer compare for the string instead of actual string comparisons.

This is assuming the thing that is compared against is _also_ a ConstString, which is not necessarily the case. 
This is probably a much longer discussion, but every time I've seen the string comparison argument in favor of using ConstString, I was never able to prove it has any measurable difference. It's obviously hard to prove a negative, but I have notes on experiments where I tried to show they make any difference inside FileSpecs, and could not prove that to be the case. I even made the StringRef::equals function extra slow (no-opt + no length shorcut) and still could not see any performance impact.


----

Getting back to the original discussion: if we feel strongly that adding this method is akin to more duplication, I don't mind closing this patch until we can do something about DWARFDeclContext and possibly remove it. 
In the meantime, the follow up patches to this can instead operate on `DWARFDeclContext::operator[]` + `Entry.Name` without too much extra burden

Reply:
> Out of curiosity, why is a pair of name + tag called a "Compiler Context"?

No terribly good reason: `DeclContext`, `DWARFDeclContext` and `CompilerDeclContext` were already taken.

Reply:
FWIW, I have no problem changing CompilerContext to use other strings if that is beneficial.

Reply:
> Out of curiosity, why is a pair of name + tag called a "Compiler Context"?

I believe this was done a long time ago for the -gmodules support where we have a context to a type and we want to find it in the right module's DWARF. So I didn't name it, I just ended up using it for the type lookups. Fine to rename if we have a better name.

> > We could make a using directive for std::vector so we don't have to mention the std::vector all of the time :
> 
> A DWARFDeclContext has a couple of functions that are used elsewhere, so we can't replace it with a vector 1:1. We'd probably still need an abstraction or a few free functions. Also [personal opinion] it is better to have `std::vector` if the abstraction _is_ just vector.

I agree. Maybe just making this a class so it can have some methods might be nice.

> 
> > comparing many matching entries becomes a pointer compare for the string instead of actual string comparisons.
> 
> This is assuming the thing that is compared against is _also_ a ConstString, which is not necessarily the case. This is probably a much longer discussion, but every time I've seen the string comparison argument in favor of using ConstString, I was never able to prove it has any measurable difference. It's obviously hard to prove a negative, but I have notes on experiments where I tried to show they make any difference inside FileSpecs, and could not prove that to be the case. I even made the StringRef::equals function extra slow (no-opt + no length shorcut) and still could not see any performance impact.

Setting a breakpoint by full path on a large project might show some speed differences, but maybe not if StringRef::operator==() is efficient enough to just look at the lengths first so there may be fewer string compares...

For FileSpec objects, I like not having to worry about the lifetime of the string. Switching FileSpec objects over to use StringRef could cause crashes if the object file that owns the strings for the FileSpec gets unloaded. We would need to switch them to use std::string as there are append path component and other things that modify these paths.

> Getting back to the original discussion: if we feel strongly that adding this method is akin to more duplication, I don't mind closing this patch until we can do something about DWARFDeclContext and possibly remove it. In the meantime, the follow up patches to this can instead operate on `DWARFDeclContext::operator[]` + `Entry.Name` without too much extra burden

It would be nice if we had one way to compare contexts that is similar across all symbol file plug-ins, so for me the `std::vector<CompilerContext>` would be nice to use everywhere. It will stop people in SymbolFile plug-ins from having to make a new similar class say for the PDB plug-in. If everyone can use `std::vector<CompilerContext>`, or if we want to make a class that contains `std::vector<CompilerContext>` and has a small API, then we can control things a bit better. Each symbol file plug-in would need to be able to make a `std::vector<CompilerContext>` from items in the debug info and then use a common API. If we make a class, we can control the comparison in methods where we want an exact match or a partial match.

Let me know if this creates a burden or if you think this is the wrong approach

Reply:
> No terribly good reason: DeclContext, DWARFDeclContext and CompilerDeclContext were already taken.

> I believe this was done a long time ago for the -gmodules support where we have a context to a type and we want to find it in the right module's DWARF.

Got it! Sounds like if we make a new common class we may have an opportunity to rename this. Depending on what we settle with, the most uninspired name might be the most appropriate: NameAndTag, or NameAndKind. Makes it pretty obvious what the class is.

> Let me know if this creates a burden or if you think this is the wrong approach

I agree having a common API is a good thing, but it feels orthogonal to what this patch was hoping to facilitate, as it would require putting IDX_Parent work on pause to go find a common interface for the existing plugins. 

This commit was here just to avoid having to write a couple of extra `Entry.Name.toStringRef`s in a subsequent patch, so I don't mind abandoning it until we have such a common interface.

> Setting a breakpoint by full path on a large project might show some speed differences, 

This is exactly what I tried, and it made no difference. Even made the paths extra lengthy.

> StringRef::operator==() is efficient enough to just look at the lengths first so there may be fewer string compares...

This is also exactly what I disabled. I tagged it with `attribute::noopt` (which also disables inlining) and also removed the shortcut that compares the lengths first, i.e., I made the comparator look at every character every time. Didn't make a dent on breakpoint setting performance. This makes me believe the opposite is true: if we got rid of ConstString, things may potentially be faster because of all the overhead in conjuring ConstStrings from other string representations (my experiments didn't replace ConstString, merely rewrote their comparator to be as slow as possible). But we are neck-deep in the program-lifetime ocean, so maybe some other day 😅